### ISSUE TITLE
Regression? Unreliable under stress

### ISSUE THREAD

**Issue by alexhultman** [warmcat/libwebsockets#545]
I'm trying to build an echo server. This is what I have so far but it behaves unreliably under stress (10k users sending messages randomly). I just want to get this server working, did I mess up somehow?

```c++
#include <libwebsockets.h>
#include <iostream>
#include <new>
using namespace std;

struct SocketExtension {
    string buffer;
};

int callback(lws *wsi, lws_callback_reasons reason, void *user, void *in, size_t len)
{
    SocketExtension *ext = (SocketExtension *) user;

    switch (reason) {

    case LWS_CALLBACK_ESTABLISHED:
    {
        new (ext) SocketExtension;
        break;
    }

    case LWS_CALLBACK_CLOSED:
    {
        ext->buffer.~string();
        break;
    }

    case LWS_CALLBACK_RECEIVE:
    {
        size_t remainingBytes = lws_remaining_packet_payload(wsi);
        if (!remainingBytes && !ext->buffer.length()) {
            lws_write(wsi, (unsigned char *) in, len, lws_frame_is_binary(wsi) ? LWS_WRITE_BINARY : LWS_WRITE_TEXT);
        } else {
            ext->buffer.append((char *) in, len);
            if (!remainingBytes) {
                lws_write(wsi, (unsigned char *) ext->buffer.data(), ext->buffer.length(), lws_frame_is_binary(wsi) ? LWS_WRITE_BINARY : LWS_WRITE_TEXT);
                ext->buffer.clear();
            }
        }
        break;
    }
    }
    return 0;
}

int main(int argc, char *argv[])
{
    lws_set_log_level(0, nullptr);

    lws_protocols protocols[] = {
        {"default", callback, sizeof(SocketExtension)},
        {nullptr, nullptr, 0}
    };

    lws_context_creation_info info = {};
    info.port = 3000;

    info.protocols = protocols;
    info.gid = info.uid = -1;
    info.options |= LWS_SERVER_OPTION_LIBEV;

    lws_context *context;
    if (!(context = lws_create_context(&info))) {
        cout << "ERR_LISTEN" << endl;
        return -1;
    }

    lws_ev_initloop(context, ev_default_loop(), 0);
    ev_run(ev_default_loop());
}


```

**Comment by lws-team** [comment]
One of the basic points of using lws is 'only send things when the connection is writable'.  You're sending things from the rx callback.

**Comment by alexhultman** [comment]
Yes I know, this was the behavior way back in the days but don't you remember we talked aobut this? You added the behavior of having the library buffer any sends that did not fit into the kernel buffer. Don't you remember I managed to send 100mb of data with using lws_write outside of the callback? Besides, I'm getting identical behavior with the stock echo server. Both this one I wrote and the one you provide as a test app behave identical.

**Comment by lws-team** [comment]
That's not the issue: you may receive n rx actions / callbacks before it can send anything...

There can be some other problem but that's a fatal problem by itself.

**Comment by alexhultman** [comment]
So basically I need to write something like what I did with node-lws: https://github.com/lpinca/node-lws/blob/master/src/lws.cpp to get a reliable echo server.

**Comment by alexhultman** [comment]
Your'e right, I managed to get it reliable now. I have updated my benchmarks table to say "User space memory scaling" which means one can figure out it will not scale this good when taking kernel space memory into account. Anyways, just wanted to say that it works now. I have updated my benchmark for lws 2.0 and the source code for the server is in benchmarks folder if you want to try and optimize it further and I can re-run the benchmarks if you want. Anyways, good night and thanks for the input.
